### PR TITLE
sink(ticdc): make the size of prep stmt cache adaptive based on the value of max_prepared_stmt_count and worker count

### LINKED ISSUE #8379: TiCDC changefeed stucks "Can't create more than max_prepared_stmt_count statements (current value: 16382)"
### What did you do?

1. create changefeed for MySQL sink
2. run workload
```
sysbench --db-driver=mysql --mysql-host=upstream-tidb.cdc-testbed-tps-1651903-1-86 --mysql-port=4000 --mysql-user=root --mysql-db=workload --tables=100 --table-size=500000 --create_secondary=off --time=3600 --threads=100 oltp_update_non_index prepare
sysbench --db-driver=mysql --mysql-host=upstream-tidb.cdc-testbed-tps-1651903-1-86 --mysql-port=4000 --mysql-user=root --mysql-db=workload --tables=100 --table-size=500000 --create_secondary=off --time=3600 --threads=100 oltp_update_non_index run
```

### What did you expect to see?

PITR changefeed should advance normal

### What did you see instead?

PITR changefeed stucks.

eed6ba7e-7d02-41f7-80d0-db3aec8ada81

In cdc log, we can see MySQL txn error: Error 1461 (42000): Can't create more than max_prepared_stmt_count statements (current value: 16382)
```
[2023/02/26 10:16:37.314 +00:00] [ERROR] [mysql.go:184] ["execute DMLs failed"] [error="[CDC:ErrReachMaxTry]reach maximu
m try: 8, error: [CDC:ErrMySQLTxnError]MySQL txn error: Error 1461 (42000): Can't create more than max_prepared_stmt_cou
nt statements (current value: 16382): [CDC:ErrMySQLTxnError]MySQL txn error: Error 1461 (42000): Can't create more than
max_prepared_stmt_count statements (current value: 16382)"] [errorVerbose="[CDC:ErrReachMaxTry]reach maximum try: 8, err
or: [CDC:ErrMySQLTxnError]MySQL txn error: Error 1461 (42000): Can't create more than max_prepared_stmt_count statements
 (current value: 16382): [CDC:ErrMySQLTxnError]MySQL txn error: Error 1461 (42000): Can't create more than max_prepared_
stmt_count statements (current value: 16382)\[ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0](http://ngithub.com/pingcap/errors.AddStack/n/tgithub.com/pingcap/errors@v0.11.5-0).
20221009092201-b66cddb77c32/errors.go:174\[ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/e](http://ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs/n/tgithub.com/pingcap/e)
rrors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\[ngithub.com/pingcap/tiflow/pkg/retry.run\n\tgithub.com/ping](http://ngithub.com/pingcap/tiflow/pkg/retry.run/n/tgithub.com/ping)
cap/tiflow/pkg/retry/retry_with_opt.go:69\[ngithub.com/pingcap/tiflow/pkg/retry.Do\n\tgithub.com/pingcap/tiflow/pkg/retry](http://ngithub.com/pingcap/tiflow/pkg/retry.Do/n/tgithub.com/pingcap/tiflow/pkg/retry)
/retry_with_opt.go:34\[ngithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql.(*mysqlBackend).execDMLWithMaxRetries\n\tgit](http://ngithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql.(*mysqlBackend).execDMLWithMaxRetries/n/tgit)
[hub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql/mysql.go:582\ngithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql](http://hub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql/mysql.go:582/ngithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql).(*m
ysqlBackend).Flush\n\[tgithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql/mysql.go:182\ngithub.com/pingcap/tiflow/cdc/s](http://tgithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/mysql/mysql.go:182/ngithub.com/pingcap/tiflow/cdc/s)
ink/dmlsink/txn.(*worker).doFlush\n\[tgithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/worker.go:214\ngithub.com/pingcap/ti](http://tgithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/worker.go:214/ngithub.com/pingcap/ti)
flow/cdc/sink/dmlsink/txn.(*worker).runBackgroundLoop.func1\n\[tgithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/worker.go](http://tgithub.com/pingcap/tiflow/cdc/sink/dmlsink/txn/worker.go):
163\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
```

### Versions of the cluster

CDC version:
[2023/02/26 10:06:56.361 +00:00] [INFO] [version.go:47] ["Welcome to Change Data Capture (CDC)"] [release-version=v6.7.0
-alpha] [git-hash=00bfd0d0580d59977adf91ea9c1e237f787d0b6c] [git-branch=heads/refs/tags/v6.7.0-alpha] [utc-build-time="2
023-02-25 11:32:40"] [go-version="go version go1.20.1 linux/amd64"] [failpoint-build=false]

### PR DESCRIPTION
**Pull request by dbsid**
### What problem does this PR solve?

Issue Number: close #8379

### What is changed and how it works?

lower the default size of `prep-stmt-cache-size` based on the default value of MySQL max_prepared_stmt_count, to make sure  when there is single one changfeed, there is no error `Error 1461 (42000): Can't create more than max_prepared_stmt_cou
nt statements (current value: 16382)`

```
	// defaultPrepStmtCacheSize is the default size of prepared statement cache
	// default size = (default max_prepared_stmt_count) / (default worker count + 1)
	// 16382 / (16 + 1) = 963
	defaultPrepStmtCacheSize = 963
```

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: pkg/sink/mysql/config.go [hunk 1]
```diff
@@ -75,10 +75,12 @@ const (
 
 	// defaultcachePrepStmts is the default value of cachePrepStmts
 	defaultCachePrepStmts = true
-	// defaultStmtCacheSize is the default size of prepared statement cache
-	defaultPrepStmtCacheSize = 10000
+	// defaultPrepStmtCacheSize is the default size of prepared statement cache
+	// default size = (default max_prepared_stmt_count) / (default worker count + 1)
+	// 16382 / (16 + 1) = 963
+	defaultPrepStmtCacheSize = 963
```
Comment (amyangfei) [inline pkg/sink/mysql/config.go:81]: How does `worker count` affect the prepare statement cache size, what happens if user sets a larger `worker-count` in sink-uri.

Comment (dbsid) [reply pkg/sink/mysql/config.go:81]: If the work-count increases, the number of prepared statements on the MySQL server will also increase, which may result in hitting the limit. The default size is set to ensure that there are no errors with a single changefeed and default work-count. However, if the user needs to increase the worker count or prepared statement cache size, they should also increase the max_prepared_stmt_count on the MySQL server. It's worth noting that TiDB does not have such a limit by default.

Comment (amyangfei) [reply pkg/sink/mysql/config.go:81]: `worker-count` is often set to a larger value, the risk of this error exists if `max_prepared_stmt_count` is not increased in TiDB/MySQL server.
Use a default value maybe not an ideal solution, is it possible to query `max_prepared_stmt_count` from downstream database and set a proper value of `prep-stmt-cache-size` based on the `max_prepared_stmt_count` and `worker-count`.

Comment (dbsid) [reply pkg/sink/mysql/config.go:81]: hmm, make sense, I've give it a try.

**Review by amyangfei — APPROVED** [review]

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 2]
```diff
@@ -68,6 +68,8 @@ type mysqlBackend struct {
 	metricTxnSinkDMLBatchCallback prometheus.Observer
 	// implement stmtCache to improve performance, especially when the downstream is TiDB
 	stmtCache *lru.Cache
+	// Indicate if the CachePrepStmts should be enabled or not
+	cachePrepStmts bool
```
Comment (hi-rustin) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:72]: Can we just access it by `cfg.CachePrepStmts`? Because we already have the cfg in this struct.

**Review by hi-rustin — COMMENTED** [review]
Thanks!

File: cdc/sink/dmlsink/txn/mysql/mysql.go (hunk 2 above)
Comment (dbsid) [reply cdc/sink/dmlsink/txn/mysql/mysql.go:72]: when on serverside, max_prepared_stmt_count is 0 or int(maxPreparedStmtCount/(cfg.WorkerCount+1)) is 0, even cfg.CachePrepStmts is true, we still need to disable prep stmt cache. So a new field cachePrepStmts is added to control if prep stmt cache should be used or not.

**Review by hi-rustin — APPROVED** [review]

File: cdc/sink/dmlsink/txn/mysql/mysql.go (hunk 2 above)
Comment (dbsid) [reply cdc/sink/dmlsink/txn/mysql/mysql.go:72]: I'm not sure if the value of cfg.CachePrepStmts will get reset or not, when the downstream MySQL config is changed and CDC need to rebuild the connection pool, or the sink URL options and the worker count are updated. So for safe a new field cachePrepStmts is added.

**Pull request merged**